### ISSUE TITLE
ensure we allow arbitrary byte sequences in path names

### ISSUE THREAD

**Issue by schroejahr2** [fish-shell/fish-shell#3989]
fish, version 2.5.0 installed via brew 
os: OSX El Captain
Terminal: xterm-256color

typing ls(space)(tab) results in crash. Freshly installed, no plugins.

crashes also with other commands (not just ls).

This happens just in my Downloads folder (which is quite stuffed)

```
❰schroejahr❙~❱✔≻ cd Downloads/                                                                               18:34:46
❰schroejahr❙~/Downloads❱✔≻ ls libc++abi.dylib: terminating with uncaught exception of type std::length_error: basic_string

[Prozess beendet]
```

**Comment by krader1961** [comment]
I can't reproduce on macOS. There is undoubtedly a bug since fish should never fail in that manner. But we're going to need more info from you. What does `locale` output? Does this happen in any other directory? Preferably one with a modest number of files. Please do `ls -l >ls.txt` in a directory where this happens and attach the `ls.txt` file to this issue. If you have the `lldb` command installed please run `ulimit -c unlimited` then reproduce the problem to get a core file. Then `lldb -c /cores/*` followed by `bt -a`.

**Comment by dutragustavo** [comment]
I detected the same crash in Ubuntu 12.04, fish version 2.0.0. 

I tried to run it with GDB to get a backtrace of the error and got this:
 
```
Program received signal SIGTTOU, Stopped (tty output).
0x00007ffff73ad1e7 in ioctl () at ../sysdeps/unix/syscall-template.S:81
81	../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) bt full
#0  0x00007ffff73ad1e7 in ioctl () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff73aca29 in tcsetpgrp (fd=<optimized out>, pgrp_id=12287) at ../sysdeps/unix/bsd/tcsetpgrp.c:29
No locals.
#2  0x000000000041f023 in main ()
No symbol table info available.
```

**Comment by schroejahr2** [comment]
`❰schroejahr❙~❱✔≻ locale                                                 00:26:48
LANG="de_DE.UTF-8"
LC_COLLATE="de_DE.UTF-8"
LC_CTYPE="de_DE.UTF-8"
LC_MESSAGES="de_DE.UTF-8"
LC_MONETARY="de_DE.UTF-8"
LC_NUMERIC="de_DE.UTF-8"
LC_TIME="de_DE.UTF-8"
LC_ALL=`

No, i did not find another directory where this happens. 

`❰schroejahr❙~/Downloads❱✘≻ ls | wc -l                                   00:28:04
     406
`
I dont want post my whole download history online. so i send the file via pm to you. feel free to forward it to devs.

will install lldb and send the files requested tomorrow.

**Comment by krader1961** [comment]
Note that lldb is part of the Xcode developer tools. You can't use gdb because fish is built with clang, not gcc, on macOS. With the `ls` output you sent me I can reproduce the failure. So don't bother installing Xcode just for this issue. Now that I can reproduce the failure we've got what we need.

**Comment by krader1961** [comment]
Also, I can reproduce this with fish built from git head so this isn't something we've fixed since 2.5.0 was released.

**Comment by krader1961** [comment]
The problem is you have an invalid file name: `A\u0303\x9cberweisung2.jpg`. Note the invalid \x9C character after the \u0303 (combining tilde). We'll need to make fish deal better with invalid UTF-8 sequences.

**Comment by krader1961** [comment]
Note that the failure is a string exception thrown from `pager_t::completion_print_item`. The question is whether we should even attempt to harden that code. I vote no since we shouldn't allow invalid UTF-8 to even make it into fish data structures. That is, we should be converting invalid UTF-8 to something valid when we convert such sequences to wide chars. Of course that means you won't be able to complete such file names unless we use some sort of round-trip meta encoding that preserves invalid UTF-8 chars. But since fish is not the proper tool for handling arbitrary byte sequences, and doing so would greatly complicate its internals I vote we simply replace invalid chars with `#define REPLACEMENT_WCHAR 0xFFFDu` when reading strings.

**Comment by krader1961** [comment]
I changed the milestone because this problem has been present for many years. There is no reason we should predicate the fish 2.6.0 release on fixing this issue.

**Comment by faho** [comment]
>I detected the same crash in Ubuntu 12.04, fish version 2.0.0.

@dutragustavo: It's not the same, and it's not a crash - fish there just receives a SIGGTOU because it's not allowed to write to that terminal because it's in the wrong process group. There have been some changes in that area, so you'll want to upgrade. You'll want to upgrade regardless, because fish has changed quite a bit in the meantime.

-----------------

So, regarding the issue at hand...

I've seen that particular encoding error before ("A~" instead of "Ü"). I think it's a filename that was encoded either as latin1 or UTF-16. Which is quite common for files you receive from windows, since that's what windows uses.

>That is, we should be converting invalid UTF-8 to something valid when we convert such sequences to wide chars. Of course that means you won't be able to complete such file names unless we use some sort of round-trip meta encoding that preserves invalid UTF-8 chars.

I believe it's worth improving this, since it's something that'll actually happen to people, and because it's more _correct_ - filenames in UNIX are arbitrary byte sequences that don't have to adhere to any particular encoding (though in practice it needs to be ASCII-compatible in that it can't have 0x2f - "/" in ASCII - or 0x00 - the string terminator, so UTF-8 is the obvious choice).

Would it be possible to convert this to the "\x" notation? I _think_ I've seen that happen, so we might already have something for it in place.

**Comment by dutragustavo** [comment]
Thanks @faho , i will upgrade my fish-shell.

**Comment by krader1961** [comment]
The directory contained a *$RECYCLE.BIN* directory which is a sure sign the user is running MS Windows inside a VMware Fusion virtual machine and sharing the directory with the guest OS. So, yes, it is quite likely the problematic file name originated from MS Windows.

As @faho reminded me path names are a special case in that they don't have to, although in practice almost always do, conform to a particular encoding. The problem is that when dealing with network file systems (e.g., NFS) or virtual machines with shared directories there is no guarantee the encoding used by a system that created the file name is the same as that of the system using the file name. So you might have a system that is using a ISO 8859 encoding such as latin-1 sharing files with a system using UTF-8 (which is probably what is happening in this issue). So, yes, we do need to treat path names as arbitrary byte sequences. That does not mean we need to, or should, treat things like scripts as arbitrary byte sequences.

**Comment by schroejahr2** [comment]
I have investigated the file. 
You are righ. There is an VM Fusion containing Windows 7 on this Machine.

It looks like ok in the Terminal, but i cant paste it here since then everysing meesses up.

But i can Autocomplete this file without a crash

https://www.dropbox.com/s/2gca6mxq3f72g59/Screenshot%202017-04-30%2000.55.38.png?dl=0

But the file is listed multiple times even it is only once in the folder.
i do not understand this behavior, but i did not write any code on this. So all i can do is to show you what happens.

The content of the file is a screenshot of a part of a website, but cutted. I am very sure but i cant guarantee -  this file was not created in windows. Since ist is a screenshot of an online wire transfer which i havent done anytime on the vm.

**Comment by faho** [comment]
>But the file is listed multiple times even it is only once in the folder.
i do not understand this behavior, but i did not write any code on this.

Glitches like that usually happen when fish and the terminal disagree on how wide a character is (often with suggestions, since that involves a bunch of trickery behind the scenes). In this case, there's something in that filename that _isn't_ a valid character, and also two codepoints that combine to make one character - that "Ã" (not "Ä" - look closely) consists of a capital A and a "combining tilde". The capital A has the width of a normal latin character, while the combining tilde isn't displayed by itself (i.e. it has zero-width).

If you would like to find out if it is the former (the invalid junk) or the latter (combining character), you could try `echo A\U0303`, then copy the resulting "Ã" and paste it to execute `echo Ã` to get it into the suggestions. If that still glitches out, please file a separate issue (though it is still quite possible that your terminal is at fault).

If it doesn't, then the junk causes the issue, which would be fixed by fixing this issue - we should never print it directly, instead encoding it into the \xYZ notation or similar.

**Comment by schroejahr2** [comment]
https://www.dropbox.com/s/kb6w8z45bsycobq/Screenshot%202017-04-30%2001.35.00.png?dl=0
looks wired

**Comment by faho** [comment]
@schroejahr2: That "ls" suggestion is still the filename, which still includes the junk!

Do this:

```fish
$ echo A\U0303
Ã # copy this
$ echo ÃbababababaÃ # paste the 'Ã' in there
$ echo # this should now suggest ÃbababababaÃ
```

if that `echo` also displays the weirdness, then there's an additional issue.

**Comment by schroejahr2** [comment]
this is what i have done

**Comment by faho** [comment]
@schroejahr2: Judging from your latest screenshot you haven't. The difference is that you've continued to use `ls`, which would suggest the _filename_, which also includes that bogus junk. You need to use another command (e.g. `echo`) so that it suggests the "Ã" in a different context.

@krader1961: I cannot reproduce this crash on archlinux with the same locale. I've used `touch (printf 'A\u0303\x9cberweisung2.jpg')` to try to recreate the offending filename, and tab-completing `ls ` results in `ls Ã\X9cberweisung2.jpg`. Have I done anything wrong or is there a difference in the underlying system libraries? Or does HFS+ weird normalization have anything to do with this?

**Comment by krader1961** [comment]
I see the tool I've been using to tell me about UTF-8 sequences has a bug. The problem file name actually contains \u009C not \x9C. In an empty directory do `touch A\u0303\u009cB '$RECYCLE.BIN'`.  Now type `echo ` and press [tab]. Boom. Without the `$RECYCLE.BIN` file the failure doesn't occur. Not sure why. I get essentially the same error message on both macOS and Ubuntu. This is the Linux failure message:

```
$ echo terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_create
fish: 'fish' terminated by signal SIGABRT (Abort)
```

Note that \u009C is not a normal character; it's the "string terminator" psuedo-char: https://codepoints.net/U+009C?lang=en. Googling didn't return any good descriptions of the purpose of this code point and how to deal with it.

**Comment by antoyo** [comment]
I've got a similar issue:

```
$ fish --version
fish, version 2.5.0
```

I use ArchLinux and the terminal is `termite`.

```
$ uname -a
Linux hostname 4.11.3-1-ARCH #1 SMP PREEMPT Sun May 28 10:40:17 CEST 2017 x86_64 GNU/Linux
```
```
$ echo $TERM
xterm-termite
```

When trying without customizations, I see the following error:
```
fish: src/pager.cpp :161 : line_t pager_t::completion_print_item(const wcstring&, const pager_t::comp_t*, size_t, size_t, size_t, bool, bool, page_rendering_t*) const:  l'assertion « desc_remaining >= 2 » a échoué.ta/\ E\cb\ca.pstate      …est/data/\ \Xc5X\cb.pstate        
fish: 'sh -c 'env HOME=$(mktemp -d) fi…' terminated by signal SIGABRT (Abandon)
```

** How to reproduce this issue: **

I type `rm -Rf directory/<Tab><Tab>` and this directory contains files with invalid characters.

Unrelated, but fish is also unable to do proper `*` expansion for some of these files:
running `rm -f *` in this directory didn't delete all the files while in `bash`, it did.

**Comment by faho** [comment]
>In an empty directory do touch A\u0303\u009cB '$RECYCLE.BIN'. Now type echo and press [tab]. Boom. Without the $RECYCLE.BIN file the failure doesn't occur. Not sure why.

@krader1961: I believe I know why - the crash seems to happen in `pager_t::completion_print_item`, but only in the else-path to `if (c->desc_width > 0 && desc_remaining > 4) {
`, i.e. the one taken when the entry is too long. You can also reproduce this by creating an invalid file and _another_ with a long (but valid) name.

This can be fixed by changing it to:

```fish
    } else if (desc_remaining > 0) {
        // No description, or it won't fit. Just add spaces.
        print_max(wcstring(desc_remaining, L' '), 0, desc_remaining, false, &line_data);
    }
```

Obviously the real fix is to get the width to be reported properly - this won't line up properly - but it's better than crashing.

**Comment by faho** [comment]
This no longer crashes, so I'm closing.

**Comment by zanchey** [comment]
Fixed as far back as 2.7, though the entry is just dropped from the list entirely.
